server/scheduler.go

```diff
@@ -297,24 +298,42 @@ func scheduleRemovePeer(cluster *clusterInfo, s Selector, filters ...Filter) (*r
 
 // scheduleTransferLeader schedules a region to transfer leader to the peer.
 func scheduleTransferLeader(cluster *clusterInfo, s Selector, filters ...Filter) (*regionInfo, *metapb.Peer) {
-	sourceStores := cluster.getStores()
-
-	source := s.SelectSource(sourceStores, filters...)
-	if source == nil {
+	stores := cluster.getStores()
+	if len(stores) == 0 {
 		return nil, nil
 	}
 
-	region := cluster.randLeaderRegion(source.GetId())
-	if region == nil {
-		return nil, nil
+	var averageLeader float64
+	for _, s := range stores {
+		averageLeader += float64(cluster.getStoreLeaderCount(s.GetId())) / float64(len(stores))
 	}
 
-	targetStores := cluster.getFollowerStores(region)
+	mostLeaderStore := s.SelectSource(stores, filters...)
+	leastLeaderStore := s.SelectTarget(stores, filters...)
 
-	target := s.SelectTarget(targetStores)
-	if target == nil {
+	if mostLeaderStore == nil && leastLeaderStore == nil {
 		return nil, nil
 	}
 
-	return region, region.GetStorePeer(target.GetId())
+	if leastLeaderStore == nil || math.Abs(mostLeaderStore.leaderScore()-averageLeader) > math.Abs(leastLeaderStore.leaderScore()-averageLeader) {
```

Comment:
I have already checked if both `leastLeaderStore` and `mostLeaderstore` are nil.